### ISSUE TITLE
TODO list

### ISSUE THREAD

**Issue by aquynh** [unicorn-engine/unicorn#354]
see the list of features/fixes we want to see in the future version of Unicorn at our **TODO** list at https://github.com/unicorn-engine/unicorn/wiki/TODO.

feel free to add what you think is important in the above Wiki page.

**Comment by farmdve** [comment]
On Windows, in uc.c, what are the reasons for this?

```c
#ifndef _WIN32
    free(uc->l1_map);
#endif
```

Doesn't this introduce a memory leak on Windows? DrMemory at least tells me so.

**Comment by aquynh** [comment]
you are right on the call to `free()`. can you send a pull request to remove those `ifndef` & `endif`?

yes at the moment memleaks are mostly due to Qemu objects. this is really a release blocker for the next version, so i put it at the top of the TODO list.

**Comment by farmdve** [comment]
I worked for a few hours and managed to bring down the leaks to about 3.

Despite my best efforts though, for some reason this region here isn't freed.

memory_region_init(uc, uc->system_memory, NULL, "system", UINT64_MAX);

The problem is again, the object system adds a child property called system[0] but when I do 

```c
OBJECT(uc->system_memory)->ref = 1;
object_unref(uc, OBJECT(uc->system_memory));
```

Even though all it's properties should be removed, the last one added, system[0] is nowhere to be found and thus isn't freed.

**Comment by aquynh** [comment]
i havent looked too close at how the object handle memory allocation yet, but it is highly likely that we have to extend it to properly free all the memory.

**Comment by farmdve** [comment]
It should work, it's just that something replaces the object pointer for uc->system_memory.

**Comment by farmdve** [comment]
I did it, I've fixed all the leaks, there are some handle leaks, but I believe this is better than before.

Dr. Memory version 1.9.0 build 4 built on Oct  2 2015 13:13:14
Dr. Memory results for pid 8628: "hook.exe"
Application cmdline: "C:\MinGW\msys64\projects\unicorn\samples\hook.exe"
Recorded 116 suppression(s) from default C:\Users\<user>\Desktop\DrMemory-Windows-1.9.0-4\bin\suppress-default.txt

Dr. Memory version 1.9.0 build 4 built on Oct  2 2015 13:13:14
Dr. Memory results for pid 8628: "hook.exe"
Application cmdline: "C:\MinGW\msys64\projects\unicorn\samples\hook.exe"
Recorded 116 suppression(s) from default C:\Users\<user>\Desktop\DrMemory-Windows-1.9.0-4\bin\suppress-default.txt

Dr. Memory version 1.9.0 build 4 built on Oct  2 2015 13:13:14
Dr. Memory results for pid 8628: "hook.exe"
Application cmdline: "C:\MinGW\msys64\projects\unicorn\samples\hook.exe"
Recorded 116 suppression(s) from default C:\Users\<user>\Desktop\DrMemory-Windows-1.9.0-4\bin\suppress-default.txt

Error 1: UNINITIALIZED READ: reading register eax
= 0 disas_insn                                          [C:/MinGW/msys64/projects/unicorn2/qemu/target-i386/translate.c:4755]
= 1 gen_intermediate_code_internal_x86_64               [C:/MinGW/msys64/projects/unicorn2/qemu/target-i386/translate.c:8513]
= 2 gen_intermediate_code_x86_64                        [C:/MinGW/msys64/projects/unicorn2/qemu/target-i386/translate.c:8562]
= 3 cpu_x86_gen_code                                    [C:/MinGW/msys64/projects/unicorn2/qemu/translate-all.c:179]
= 4 tb_gen_code_x86_64                                  [C:/MinGW/msys64/projects/unicorn2/qemu/translate-all.c:1114]
= 5 tb_find_slow_x86_64                                 [C:/MinGW/msys64/projects/unicorn2/qemu/cpu-exec.c:365]
= 6 tb_find_fast_x86_64                                 [C:/MinGW/msys64/projects/unicorn2/qemu/cpu-exec.c:393]
= 7 cpu_x86_exec                                        [C:/MinGW/msys64/projects/unicorn2/qemu/cpu-exec.c:201]
= 8 tcg_cpu_exec_x86_64                                 [C:/MinGW/msys64/projects/unicorn2/qemu/cpus.c:214]
= 9 tcg_exec_all_x86_64                                 [C:/MinGW/msys64/projects/unicorn2/qemu/cpus.c:234]
= 10 qemu_tcg_cpu_thread_fn_x86_64                       [C:/MinGW/msys64/projects/unicorn2/qemu/cpus.c:160]
= 11 win32_start_routine@4                               [util/qemu-thread-win32.c:282]
= 12 msvcrt.dll!_endthreadex
= 13 msvcrt.dll!_endthreadex
= 14 KERNEL32.dll!BaseThreadInitThunk                   +0x11     (0x75bc337a <KERNEL32.dll+0x1337a>)
Note: @0:00:00.798 in thread 11796
Note: instruction: test   %eax %eax

===========================================================================
FINAL SUMMARY:

DUPLICATE ERROR COUNTS:
	Error    1:      3

SUPPRESSIONS USED:

ERRORS FOUND:
      0 unique,     0 total unaddressable access(es)
      1 unique,     3 total uninitialized access(es)
      0 unique,     0 total invalid heap argument(s)
      0 unique,     0 total GDI usage error(s)
      0 unique,     0 total handle leak(s)
      0 unique,     0 total warning(s)
      0 unique,     0 total,      0 byte(s) of leak(s)
      0 unique,     0 total,      0 byte(s) of possible leak(s)
ERRORS IGNORED:
      6 potential error(s) (suspected false positives)
         (details: C:\Users\<user>\Desktop\DrMemory-Windows-1.9.0-4\drmemory\logs\DrMemory-hook.exe.8628.000\potential_errors.txt)
      3 unique,     3 total,     90 byte(s) of still-reachable allocation(s)
         (re-run with "-show_reachable" for details)
Details: C:\Users\<user>\Desktop\DrMemory-Windows-1.9.0-4\drmemory\logs\DrMemory-hook.exe.8628.000\results.txt

**Comment by aquynh** [comment]
awesome, looks forward to your PR to see what Valgrind has to say!

regarding the "UNINITIALIZED READ", exactly where the bug is in source? according to the reported output above, it seems you modified `qemu/target-i386/translate.c`, so i cannot find that place. but this issue should be easy to fix.

**Comment by farmdve** [comment]
My particular build contains the post instruction hook. But this seems to be from
```c
    // end address tells us to stop emulation
    if (s->pc == s->uc->addr_end) {
```

Ths line. Basically s->pc is used before it's set to s->pc = pc_start;

**Comment by aquynh** [comment]
oops that is an important bug. just pushed a commit to fix this uninitialized read, thanks.

**Comment by aquynh** [comment]
@farmdve, can you send me an email so i can ask you few things?

thanks

**Comment by farmdve** [comment]
A small patch is located here. https://github.com/farmdve/unicorn/commit/0585ce1a4a89a0253973e739894a1d0479b448aa

Here is the output of Valgrind under Linux with my patch http://pastebin.com/cLkx4aLt

Let me explain why it still shows leaks. Blocks 1 through 13 except block 10(tcg_register_jit) are quirks of GLIB, it's memory that cannot be destroyed no matter how many times we attempt to free it. It seems that GLIB considers it faster to just let the OS reclaim it. I spent 50+hrs trying to fix something that needs to be fixed from GLIB's side.

Block 10 is a real leak though, I have not fixed it because if it turns out it's unused(unless Usercorn uses it?) we should remove the code altogether. It seems to be part of some gdb interface.

**Comment by aquynh** [comment]
the Pastebin is a bit confused: exactly where is the block 5?

**Comment by farmdve** [comment]
Block 5

> ==11223== 842 bytes in 1 blocks are still reachable in loss record 5 of 6
> ==11223==    at 0x53F8C1D: malloc (vg_replace_malloc.c:299)
> ==11223==    by 0x5652A38: g_malloc (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.3200.4)
> ==11223==    by 0x4D7505: tcg_register_jit_int_x86_64 (tcg.c:2889)
> ==11223==    by 0x4D2459: tcg_register_jit_x86_64 (tcg-target.c:2407)
> ==11223==    by 0x4BF164: tcg_exec_init_x86_64 (translate-all.c:726)
> ==11223==    by 0x45CAC0: tcg_init (accel.c:46)
> ==11223==    by 0x45CD13: accel_init_machine (accel.c:116)
> ==11223==    by 0x45CB05: configure_accelerator (accel.c:73)
> ==11223==    by 0x45E293: machine_initialize (vl.c:126)
> ==11223==    by 0x403A34: uc_open (uc.c:240)

Leak is on this line https://github.com/unicorn-engine/unicorn/blob/master/qemu/tcg/tcg.c#L2889

**Comment by aquynh** [comment]
these code are unused, and i just removed them in the commit https://github.com/unicorn-engine/unicorn/commit/e0cb02569e7caf5ca09b647d6e27dff54ee0eec3, so this problem is solved

**Comment by lunixbochs** [comment]
Try running with `G_DEBUG=always-malloc` to clean up the glib valgrind output a bit.

**Comment by farmdve** [comment]
@lunixbochs 

I've already done that. The last leaks shown with G_DEBUG=always-malloc and G_SLICE=gc-friendly are quirks of the GLIB implementation. They cannot be fixed and thus are not leaks at all.

For me with my changes, there are no other leaks reported, this of course excludes other arches, like I mentioned some or all other arches lack release interfaces and thus lack some memory cleanup.

**Comment by aquynh** [comment]
@farmdve, that last item was created by @McLovi9, and i have no idea about it

**Comment by aquynh** [comment]
@farmdve, can you open a new issue for memleak, or just move the discussion to the issue https://github.com/unicorn-engine/unicorn/issues/258?

**Comment by farmdve** [comment]
I see that performance optimization in hooks is added.

Here are some that I've personally found to be good optimizations.

uc_reg_read/write currently have a severe penalty as they call X86_CPU(uc, mycpu) every single time. Calling this is expensive.

The improvement I added on my own dev branch is that I added a single line 

CPUX86State *env = mycpu->env_ptr;

And then I just do 
```c
                case UC_X86_REG_EAX:
                    *(int32_t *)value = env->regs[R_EAX];
                    break;
```

 for each register. We eliminate X86_CPU and gain lots of speed.
